### ISSUE TITLE
resource_keycloak_oidc_identity_provider does not support setting up a Google IDP

### ISSUE THREAD

**Issue by tomrutsaert** [mrparkers/terraform-provider-keycloak#231]
It is not possible to set up a google IDP with the resource_keycloak_oidc_identity_provider.

The _kc oidc idp resource_ assumes certain things that are not valid for most Identity Provider implementations that extend from IdentityProvider.java interface in the keycloak code base.
Example: GoogleIdentityProvider.java

In case of the _google Idp_ I found that:
============================
- authorization_url: is required by the _oidc idp resource_, but not used by _google idp_ in UI
- token_url: is required by the _oidc idp resource_, but not used by _google idp_ in UI
- user_info_url: is required by the _oidc idp resource_, but not used by _google idp_ in UI, It does need to be set to "https://openidconnect.googleapis.com/v1/userinfo" otherwise GoogleIdentityProvider code will not work as expected. + You need to set it, as _oidc idp resource_ uses this field to determine the value of disableUserInfo. For the google idp you can set this field independently.
- jwks_url: this field needs to be set, If you want to set _use_jwks_url to be true. But I google idp does not need to set a jwks_url parameter.
- Many other fields have default values, where google idp does not expect them.
For some of these fields I did not find a workaround with as result that: I can not make my _google idp_ config work through terraform.

Possible solutions:
==============
1. make a separate _google idp resource_. (downside: someone else still need to make one for each social provider)
2. Make the generic idp resource more lax. 
     - Do not assume settings based on other settings, ex: use_jwks_url and jwks__url
     - Remove all default values(except provider_id)
     - Remove all required fields(except for realm, provider_id, and alias)
(downside: backwards compatibility for people already using this resource and expect the default values)

@mrparkers and @AndrewChubatiuk, What are your opinions?
Would it best to make a PR for solution 1 or 2?

@nagytech I saw in another issue(#141) your were setting u a google idp, Did you got it working, if yes how? Do you have an example?

**Comment by mrparkers** [comment]
I don't really have any experience using this particular feature of Keycloak, but generally I am in favor of having separate resources for this so we can be more explicit.  I agree that making changes to the generic idp resource might introduce changes that affect current users, so I like the idea of making a separate google idp resource.

**Comment by nagytech** [comment]
@tomrutsaert 

I ended up re-using the `resource_keycloak_oidc_identity_provider.go` resource, but had to make a few modifications along the way.  If I recall correctly, I made 'authorization_url' and 'token_url'  optional since Keycloak already knows the correct values - and I think Keycloak rejected any dummy values I threw at it.  Additionally, I needed to add a 'client_secret' to the identity provider data fields and possibly some other minor changes.

At the end of the day, there were a lot of other properties I wanted to set in Keycloak which this project didn't support without extending it.  For instance, setting the default identity provider for a given client, setting the allowed redirect urls, and using custom login flows, etc.  Given that my solution would have been half in Terraform and half in configuration scripts, I just opted for using `kcadm.sh` for the initial system configuration.  And, then it just seemed that each realm would be easier managed by engineers using Keycloak's import/export functionality.  

I wish I had more time to help extend this project, but I was fairly new to Terraform and Go and ended up feeling other pressures to get things done.  If I get some inspiration + free time in the future I'll come back to this and try again.

**Comment by tomrutsaert** [comment]
@nagytech  Thanks for the information.
Just heads up, in the mean time custom flows is a supported feature.
Also, keep in mind that on realm level(atrtibutes) and on idp level(extra_config) it is possible to set any attribute that is not (yet) supported in the this provider. 

for example, frontendUrl is new field and not supported by the provider, but you can do:

```
resource "keycloak_realm" "master" {
  realm                = "master"
  enabled              = true
 
  attributes = {
    frontendUrl = "https://keycloak.mycompany.com/auth/"
  }
```
Perhaps it would be good idea to also add this type of mechanism to client resources

@mrparkers, 
I will take a look if I can make a separate google idp, I might need do some copy paste work, thus that means perhaps duplicate code.....

**Comment by mrparkers** [comment]
Closed via #240
